File: git_visual.py
# git_visual.py
import sys
import argparse
from git import Repo, Commit, Actor
from datetime import datetime
import os

COLORS = {
    'red': '\033[91m',
    'green': '\033[92m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'cyan': '\033[96m',
    'reset': '\033[0m',
}

class GitVisual:
    def __init__(self, repo_path='.', color=True, graph=True, all_branches=False,
                 decorate=True, oneline=False, author=None, since=None, until=None, max_count=None):
        self.repo_path = repo_path
        self.color = color
        self.graph = graph
        self.all = all_branches
        self.decorate = decorate
        self.oneline = oneline
        self.author = author
        self.since = since
        self.until = until
        self.max_count = max_count
        self.repo = None

    def open_repo(self):
        try:
            self.repo = Repo(self.repo_path)
        except Exception as e:
            print(f"Error opening repository: {e}", file=sys.stderr)
            sys.exit(1)

    def colorize(self, text, color):
        if self.color:
            return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"
        return text

    def get_commits(self):
        # Строим опции для rev-list
        args = []
        if self.all:
            args.append('--all')
        if self.author:
            args.append(f'--author={self.author}')
        if self.since:
            args.append(f'--since={self.since}')
        if self.until:
            args.append(f'--until={self.until}')
        if self.max_count:
            args.append(f'-n {self.max_count}')
        # Получаем коммиты
        revs = self.repo.git.rev_list('--pretty=format:%h|%an|%ae|%ad|%s', *args).splitlines()
        commits = []
        for line in revs:
            if line.startswith('commit '):
                continue
            parts = line.split('|', 4)
            if len(parts) >= 5:
                h, author, email, date, msg = parts[0], parts[1], parts[2], parts[3], parts[4]
                commits.append({
                    'hash': h,
                    'author': author,
                    'email': email,
                    'date': date,
                    'message': msg
                })
        return commits

    def print_commit(self, commit, prefix='', last=True, branches=None):
        if branches is None:
            branches = {}
        connector = '└── ' if last else '├── '
        hash_color = 'yellow' if self.color else ''
        author_color = 'cyan' if self.color else ''
        date_color = 'green' if self.color else ''
        msg_color = 'reset'

        # Определяем ветки для этого коммита
        refs = branches.get(commit['hash'], [])
        ref_str = ' (' + ', '.join(refs) + ')' if refs and self.decorate else ''

        if self.oneline:
            line = f"{connector}{self.colorize(commit['hash'], hash_color)} {self.colorize(commit['message'], msg_color)}{self.colorize(ref_str, 'blue')}"
            print(prefix + line)
        else:
            print(f"{prefix}{connector}{self.colorize(commit['hash'], hash_color)}{self.colorize(ref_str, 'blue')}")
            print(f"{prefix}    {self.colorize('Author:', author_color)} {self.colorize(commit['author'], author_color)} <{commit['email']}>")
            print(f"{prefix}    {self.colorize('Date:', date_color)} {self.colorize(commit['date'], date_color)}")
            print(f"{prefix}    {self.colorize(commit['message'], msg_color)}")
            print()

    def show_graph(self):
        # Получаем коммиты
        commits = self.get_commits()
        if not commits:
            print("No commits found.")
            return
        # Собираем информацию о ветках для декора
        branches = {}
        if self.decorate:
            for ref in self.repo.references:
                if ref.is_detached:
                    continue
                try:
                    commit_hash = ref.commit.hexsha[:7]
                    branches.setdefault(commit_hash, []).append(ref.name)
                except:
                    pass
            # Добавляем HEAD
            if self.repo.head.is_detached:
                try:
                    h = self.repo.head.commit.hexsha[:7]
                    branches.setdefault(h, []).append('HEAD')
                except:
                    pass
            else:
                try:
                    h = self.repo.head.commit.hexsha[:7]
                    branches.setdefault(h, []).append('HEAD -> ' + self.repo.active_branch.name)
                except:
                    pass

        # Печатаем каждый коммит с отступами
        for i, c in enumerate(commits):
            last = (i == len(commits) - 1)
            self.print_commit(c, prefix='', last=last, branches=branches)

def main():
    parser = argparse.ArgumentParser(description="Git Visual Client")
    parser.add_argument('--graph', action='store_true', default=True, help='Show ASCII graph')
    parser.add_argument('--all', action='store_true', help='Show all branches')
    parser.add_argument('--decorate', action='store_true', default=True, help='Show refs')
    parser.add_argument('--oneline', action='store_true', help='Compact output')
    parser.add_argument('--author', help='Filter by author')
    parser.add_argument('--since', help='Commits after date')
    parser.add_argument('--until', help='Commits before date')
    parser.add_argument('-n', '--max-count', type=int, help='Limit number of commits')
    parser.add_argument('--color', action='store_true', default=True, help='Color output')
    parser.add_argument('--path', default='.', help='Repository path')
    args = parser.parse_args()

    visual = GitVisual(
        repo_path=args.path,
        color=args.color,
        graph=args.graph,
        all_branches=args.all,
        decorate=args.decorate,
        oneline=args.oneline,
        author=args.author,
        since=args.since,
        until=args.until,
        max_count=args.max_count
    )
    visual.open_repo()
    visual.show_graph()

if __name__ == '__main__':
    main()

File: git_visual.go
// git_visual.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing"
	"github.com/go-git/go-git/v5/plumbing/object"
	"github.com/go-git/go-git/v5/plumbing/storer"
)

type GitVisual struct {
	repo     *git.Repository
	graph    bool
	all      bool
	decorate bool
	oneline  bool
	author   string
	since    string
	until    string
	maxCount int
	color    bool
}

func (g *GitVisual) openRepo(path string) error {
	var err error
	g.repo, err = git.PlainOpen(path)
	return err
}

func (g *GitVisual) getCommits() ([]*object.Commit, error) {
	var commits []*object.Commit
	var iter storer.RevIter
	var err error
	if g.all {
		iter, err = g.repo.Log(&git.LogOptions{All: true})
	} else {
		ref, err := g.repo.Head()
		if err != nil {
			return nil, err
		}
		iter, err = g.repo.Log(&git.LogOptions{From: ref.Hash()})
	}
	if err != nil {
		return nil, err
	}
	err = iter.ForEach(func(c *object.Commit) error {
		// Применяем фильтры
		if g.author != "" && !strings.Contains(c.Author.Name, g.author) {
			return nil
		}
		// Ограничение по дате можно добавить
		commits = append(commits, c)
		return nil
	})
	if err != nil {
		return nil, err
	}
	if g.maxCount > 0 && len(commits) > g.maxCount {
		commits = commits[:g.maxCount]
	}
	return commits, nil
}

func (g *GitVisual) getRefs() (map[string][]string, error) {
	refs := make(map[string][]string)
	// Ветки
	branches, err := g.repo.Branches()
	if err != nil {
		return nil, err
	}
	err = branches.ForEach(func(ref *plumbing.Reference) error {
		hash := ref.Hash().String()[:7]
		refs[hash] = append(refs[hash], ref.Name().Short())
		return nil
	})
	if err != nil {
		return nil, err
	}
	// Теги
	tags, err := g.repo.Tags()
	if err != nil {
		return nil, err
	}
	err = tags.ForEach(func(ref *plumbing.Reference) error {
		hash := ref.Hash().String()[:7]
		refs[hash] = append(refs[hash], ref.Name().Short())
		return nil
	})
	if err != nil {
		return nil, err
	}
	// HEAD
	head, err := g.repo.Head()
	if err == nil {
		hash := head.Hash().String()[:7]
		refs[hash] = append(refs[hash], "HEAD")
	}
	return refs, nil
}

func (g *GitVisual) colorize(text, color string) string {
	if !g.color {
		return text
	}
	colors := map[string]string{
		"red":    "\033[91m",
		"green":  "\033[92m",
		"yellow": "\033[93m",
		"blue":   "\033[94m",
		"cyan":   "\033[96m",
		"reset":  "\033[0m",
	}
	return colors[color] + text + colors["reset"]
}

func (g *GitVisual) printCommit(c *object.Commit, prefix string, last bool, refs map[string][]string) {
	connector := "└── "
	if !last {
		connector = "├── "
	}
	hashShort := c.Hash.String()[:7]
	authorName := c.Author.Name
	date := c.Author.When.Format("2006-01-02 15:04")
	msg := strings.Split(c.Message, "\n")[0]

	refStr := ""
	if g.decorate {
		if r, ok := refs[hashShort]; ok {
			refStr = " (" + strings.Join(r, ", ") + ")"
		}
	}
	if g.oneline {
		fmt.Printf("%s%s%s%s%s\n", prefix, connector,
			g.colorize(hashShort, "yellow"),
			g.colorize(refStr, "blue"),
			" "+g.colorize(msg, "reset"))
	} else {
		fmt.Printf("%s%s%s%s\n", prefix, connector, g.colorize(hashShort, "yellow"), g.colorize(refStr, "blue"))
		fmt.Printf("%s    %s %s <%s>\n", prefix, g.colorize("Author:", "cyan"), authorName, c.Author.Email)
		fmt.Printf("%s    %s %s\n", prefix, g.colorize("Date:", "green"), date)
		fmt.Printf("%s    %s\n\n", prefix, g.colorize(msg, "reset"))
	}
}

func (g *GitVisual) showGraph() error {
	commits, err := g.getCommits()
	if err != nil {
		return err
	}
	refs, err := g.getRefs()
	if err != nil {
		return err
	}
	for i, c := range commits {
		last := i == len(commits)-1
		g.printCommit(c, "", last, refs)
	}
	return nil
}

func main() {
	var path, author, since, until string
	var graph, all, decorate, oneline, color bool
	var maxCount int
	flag.StringVar(&path, "path", ".", "Repository path")
	flag.BoolVar(&graph, "graph", true, "Show graph")
	flag.BoolVar(&all, "all", false, "Show all branches")
	flag.BoolVar(&decorate, "decorate", true, "Show refs")
	flag.BoolVar(&oneline, "oneline", false, "Compact output")
	flag.StringVar(&author, "author", "", "Filter by author")
	flag.StringVar(&since, "since", "", "Commits after date")
	flag.StringVar(&until, "until", "", "Commits before date")
	flag.IntVar(&maxCount, "n", 0, "Limit number of commits")
	flag.BoolVar(&color, "color", true, "Color output")
	flag.Parse()

	v := &GitVisual{
		graph:    graph,
		all:      all,
		decorate: decorate,
		oneline:  oneline,
		author:   author,
		since:    since,
		until:    until,
		maxCount: maxCount,
		color:    color,
	}
	if err := v.openRepo(path); err != nil {
		fmt.Fprintf(os.Stderr, "Error opening repo: %v\n", err)
		os.Exit(1)
	}
	if err := v.showGraph(); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}
